File: retos-python/13_robot.py
# Resultado 5 ⭐
def is_robot_back(moves):
    final_moves = ""
    i = 0
    changes = {
        'U': 'D',
        'D': 'U',
        'R': 'L',
        'L': 'R'
    }

    while i < len(moves):
        move = moves[i]
        if move == '*':
            final_moves += moves[i + 1]
        elif move == '!':
            tmp_move = moves[i + 1]
            final_moves += changes[tmp_move]
            i += 1
        else:
            final_moves += move
        i += 1

    while '?' in final_moves:
        index = final_moves.index('?')
        values = final_moves[index:index + 2]
        temp = final_moves[:index]

        if values[1] not in temp:
            final_moves = final_moves.replace(values, values[1], 1)
        else:
            final_moves = final_moves.replace(values, '', 1)

    x, y = 0, 0

    operations = {
        'R': (0, 1),
        'L': (0, -1),
        'U': (1, 0),
        'D': (-1, 0),
    }

    for move in final_moves:
        dx, dy = operations[move]
        x += dx
        y += dy

    return True if x == 0 and y == 0 else [y, x]

File: retos-ts/13_robot.ts
// Resultado 5 ⭐
function isRobotBack(moves: string): true | [number, number] {
    let finalmoves: string = ''
    let i: number = 0
    const changes = {
        U: 'D',
        D: 'U',
        R: 'L',
        L: 'R'
    }

    while(i < moves.length){
        const move = moves[i];    
        if(move === '*'){
            finalmoves += moves[i + 1]
        }else if(move === '!'){
            const tmpmove = moves[i+1]
            finalmoves += changes[tmpmove] 
            i++         
        }else{
            finalmoves += move
        }

        i++
    }

    while(finalmoves.includes('?')){
        const index = finalmoves.indexOf('?');
        const values = finalmoves.substring(index,index+2);
        const temp = finalmoves.slice(0,index)
    
        if(!temp.includes(values[1])){
            finalmoves = finalmoves.replace(values,values[1])
        }else{
            finalmoves = finalmoves.replace(values,'')
        }
    }

    let x: number = 0
    let y: number = 0

    const operations = {
        R: () => y++,
        L: () => y--,
        U: () => x++,
        D: () => x--,
    }

    for(const m of finalmoves){
        operations[m]()
    }

    return x === 0 && y === 0 || [y , x]
}